### ISSUE TITLE
mf.update_from_chk() does nothing to a mean-field object

### ISSUE THREAD

**Issue by cychow2** [pyscf/pyscf#1318]
Hi, it seems that `mf.update_from_chk()` does nothing to a mean-field object. The calculation doesn't start from the final result of the input chkfile, which can be seen from the scf cycle energies.

Below is an example using a molecular system.
Here's a comparison between restarting a calculation from chkfile with setting `mf.init_guess`, using `mf.from_chk()` and using `mf.update_from_chk()`:

### The code that starts from a default initial guess:
```
from pyscf import gto, dft

cell = gto.M(
    atom = '''H     0.      0.      0.    
              F     0.  0.  1.72''',
    basis = 'ccpvdz',
    verbose = 4,
)

mf = dft.KS(cell).density_fit()
mf.chkfile='mol_dz.hdf5'
mf.kernel()
```

grep cycle of the output:
```
diis_start_cycle = 1
SCF max_cycles = 50
cycle= 1 E= -99.4315635429608  delta_E= 0.0712  |g|= 0.84  |ddm|= 0.758
cycle= 2 E= -98.352461368606  delta_E= 1.08  |g|= 2.12  |ddm|= 1.89
cycle= 3 E= -99.5969416697485  delta_E= -1.24  |g|= 0.332  |ddm|= 1.65
cycle= 4 E= -99.627040210629  delta_E= -0.0301  |g|= 0.164  |ddm|= 0.231
cycle= 5 E= -99.6198222028654  delta_E= 0.00722  |g|= 0.206  |ddm|= 0.106
cycle= 6 E= -99.6248190568322  delta_E= -0.005  |g|= 0.157  |ddm|= 0.0387
cycle= 7 E= -99.6313806899915  delta_E= -0.00656  |g|= 0.0035  |ddm|= 0.116
cycle= 8 E= -99.631383383932  delta_E= -2.69e-06  |g|= 0.00166  |ddm|= 0.0038
cycle= 9 E= -99.6313841461299  delta_E= -7.62e-07  |g|= 0.000193  |ddm|= 0.00137
cycle= 10 E= -99.6313841565318  delta_E= -1.04e-08  |g|= 5.28e-07  |ddm|= 0.000142
cycle= 11 E= -99.631384156532  delta_E= -2.13e-13  |g|= 1.16e-08  |ddm|= 4.89e-07
Extra cycle  E= -99.631384156532  delta_E= 7.11e-14  |g|= 5.74e-08  |ddm|= 4.56e-08
```

### Restart by setting `mf.init_guess`:
```
from pyscf import gto, dft

cell = gto.M(
    atom = '''H     0.      0.      0.    
              F     0.  0.  1.72''',
    basis = 'ccpvdz',
    verbose = 4,
)

mf = dft.KS(cell).density_fit()
mf.chkfile='mol_dz.hdf5'
mf.init_guess = 'chkfile'
mf.kernel()
```

grep cycle of the output:
```
diis_start_cycle = 1
SCF max_cycles = 50
cycle= 1 E= -99.6313841565266  delta_E= -9.95e-14  |g|= 2.89e-07  |ddm|= 2.29e-07
Extra cycle  E= -99.6313841565262  delta_E= 3.98e-13  |g|= 1.46e-06  |ddm|= 1.16e-06
```

### Restart using `mf.from_chk()`:
```
from pyscf import gto, dft

cell = gto.M(
    atom = '''H     0.      0.      0.    
              F     0.  0.  1.72''',
    basis = 'ccpvdz',
    verbose = 4,
)

mf = dft.KS(cell).density_fit()
dm = mf.from_chk("mol_dz.hdf5")
mf.kernel(dm)
```

grep cycle of the output:
```
diis_start_cycle = 1
SCF max_cycles = 50
cycle= 1 E= -99.6313841565267  delta_E= -1.28e-13  |g|= 2.89e-07  |ddm|= 2.29e-07
Extra cycle  E= -99.6313841565262  delta_E= 4.26e-13  |g|= 1.46e-06  |ddm|= 1.16e-06
```

### Restart using `mf.update_from_chk()`:
```
from pyscf import gto, dft

cell = gto.M(
    atom = '''H     0.      0.      0.    
              F     0.  0.  1.72''',
    basis = 'ccpvdz',
    verbose = 4,
)

mf = dft.KS(cell).density_fit()
mf.update_from_chk("mol_dz.hdf5")
mf.kernel()
```

grep cycle of the output:
```
diis_start_cycle = 1
SCF max_cycles = 50
cycle= 1 E= -99.4315635429608  delta_E= 0.0712  |g|= 0.84  |ddm|= 0.758
cycle= 2 E= -98.3524613686059  delta_E= 1.08  |g|= 2.12  |ddm|= 1.89
cycle= 3 E= -99.5969416697484  delta_E= -1.24  |g|= 0.332  |ddm|= 1.65
cycle= 4 E= -99.6270402106289  delta_E= -0.0301  |g|= 0.164  |ddm|= 0.231
cycle= 5 E= -99.6198222028654  delta_E= 0.00722  |g|= 0.206  |ddm|= 0.106
cycle= 6 E= -99.6248190568322  delta_E= -0.005  |g|= 0.157  |ddm|= 0.0387
cycle= 7 E= -99.6313806899915  delta_E= -0.00656  |g|= 0.0035  |ddm|= 0.116
cycle= 8 E= -99.6313833839319  delta_E= -2.69e-06  |g|= 0.00166  |ddm|= 0.0038
cycle= 9 E= -99.6313841461299  delta_E= -7.62e-07  |g|= 0.000193  |ddm|= 0.00137
cycle= 10 E= -99.631384156532  delta_E= -1.04e-08  |g|= 5.28e-07  |ddm|= 0.000142
cycle= 11 E= -99.631384156532  delta_E= 1.42e-14  |g|= 1.16e-08  |ddm|= 4.89e-07
Extra cycle  E= -99.631384156532  delta_E=    0  |g|= 5.74e-08  |ddm|= 4.56e-08
```

In other words, `mf.update_from_chk()` is the only method that fails to restart the calculation from the chkfile, and it's obviously using the default initial guess, juding from its identical output to that of the original calculation.

According to pyscf documentation, it should 'Read attributes from the chkfile then replace the attributes of current object.'. But it seems it's not working as expected?

**Comment by hebrewsnabla** [comment]
> According to pyscf documentation, it should 'Read attributes from the chkfile then replace the attributes of current object.'. But it seems it's not working as expected?

Because the density matrix is not an attribute of the mean-field object (while the mo_coeff and mo_occ is).
So you need to pass the dm to kernel
```
from pyscf import gto, dft

cell = gto.M(
    atom = '''H     0.      0.      0.    
              F     0.  0.  1.72''',
    basis = 'ccpvdz',
    verbose = 4,
)

mf = dft.KS(cell).density_fit()
mf.update_from_chk("mol_dz.hdf5")
dm = mf.make_rdm1()
mf.kernel(dm)
```
Here, `make_rdm1` will use the updated mo_coeff and mo_occ to produce the required dm.

**Comment by cychow2** [comment]
Thank you @hebrewsnabla , it works.

However, when I try to use this dm as the initial guess for a calculation with a larger basis set (tz instead of dz) as shown in the following code

```
from pyscf import gto, dft

cell = gto.M(
    atom = '''H     0.      0.      0.    
              F     0.  0.  1.72''',
    basis = 'ccpvtz',
    verbose = 4,
)

mf = dft.KS(cell).density_fit()
mf.update_from_chk("mol_dz.hdf5")
dm = mf.make_rdm1()
mf.kernel(dm)
```
I get the following error
```
Traceback (most recent call last):
  File "/Users/chow/research/test/update_from_chk_example/mol_tz_update_dz.py", line 13, in <module>
    mf.kernel(dm)
  File "/Users/chow/git clone/pyscf/pyscf/lib/misc.py", line 638, in aliased_fn
    return getattr(self, fname)(*args, **kwargs)
  File "/Users/chow/git clone/pyscf/pyscf/scf/hf.py", line 1669, in scf
    kernel(self, self.conv_tol, self.conv_tol_grad,
  File "/Users/chow/git clone/pyscf/pyscf/scf/hf.py", line 128, in kernel
    vhf = mf.get_veff(mol, dm)
  File "/Users/chow/git clone/pyscf/pyscf/dft/rks.py", line 69, in get_veff
    ks.initialize_grids(mol, dm)
  File "/Users/chow/git clone/pyscf/pyscf/dft/rks.py", line 489, in initialize_grids
    self.grids = prune_small_rho_grids_(self, self.mol, dm,
  File "/Users/chow/git clone/pyscf/pyscf/dft/rks.py", line 252, in prune_small_rho_grids_
    rho = ks._numint.get_rho(mol, dm, grids, ks.max_memory)
  File "/Users/chow/git clone/pyscf/pyscf/dft/numint.py", line 2487, in get_rho
    for ao, mask, weight, coords \
  File "/Users/chow/git clone/pyscf/pyscf/dft/numint.py", line 2529, in block_loop
    ao = self.eval_ao(mol, coords, deriv=deriv, non0tab=non0, out=buf)
  File "/Users/chow/git clone/pyscf/pyscf/dft/numint.py", line 2499, in eval_ao
    return eval_ao(mol, coords, deriv, shls_slice, non0tab, out, verbose)
  File "/Users/chow/git clone/pyscf/pyscf/dft/numint.py", line 111, in eval_ao
    return mol.eval_gto(feval, coords, comp, shls_slice, non0tab, out=out)
  File "/Users/chow/git clone/pyscf/pyscf/gto/eval_gto.py", line 113, in eval_gto
    ao = numpy.ndarray((comp,nao,ngrids), buffer=out)
TypeError: buffer is too small for requested array
```

I tried to do the same thing (projecting to a larger basis set) with another method `mf.from_chk()`, but that doesn't work for periodic systems, as I mentioned in another issue #1317 . Would you have any suggestion for achieving this?

**Comment by hebrewsnabla** [comment]
So, as you observed, `from_chk()` works but `update_from_chk()` doesn't, for the small-to-larger-basis restart? That's reasonable, because `update_from_chk()` only updates the attributes but does not perform the projection, while `from_chk()` performs the projection (so does init_guess by chkfile).

Currently I have no idea on the periodic issue.

**Comment by sunqm** [comment]
PBC SCF calculations are similar to molecular SCF. `from_chk` returns a projected density matrix while `update_from_chk` simply overwrites the scf object
